src/Type/Php/BcMathStringOrNullReturnTypeExtension.php

```diff
@@ -0,0 +1,120 @@
+use PHPStan\Type\BooleanType;
+use PHPStan\Type\Constant\ConstantStringType;
+use PHPStan\Type\NullType;
+use PHPStan\Type\StringType;
+use PHPStan\Type\Type;
+use PHPStan\Type\UnionType;
+use function in_array;
+
+class BcMathStringOrNullReturnTypeExtension implements \PHPStan\Type\DynamicFunctionReturnTypeExtension
+{
+
+	public function isFunctionSupported(FunctionReflection $functionReflection): bool
+	{
+		return in_array($functionReflection->getName(), ['bcdiv', 'bcmod', 'bcpowmod', 'bcsqrt'], true);
+	}
+
+	public function getTypeFromFunctionCall(FunctionReflection $functionReflection, FuncCall $functionCall, Scope $scope): Type
+	{
+		if ($functionReflection->getName() === 'bcsqrt') {
+			return $this->getTypeForBcSqrt($functionCall, $scope);
+		}
+
+		if ($functionReflection->getName() === 'bcpowmod') {
+			return $this->getTypeForBcPowMod($functionCall, $scope);
+		}
+
+		$secondArgument = $scope->getType($functionCall->args[1]->value);
+		$defaultReturnType = new UnionType([new StringType(), new NullType()]);
+
+		if (isset($functionCall->args[1]) === false) {
+			return $defaultReturnType;
+		}
+
+		if (($secondArgument instanceof ConstantStringType && $secondArgument->getValue() === '0')) {
+			return new NullType();
+		}
+
+		if ($secondArgument instanceof ConstantStringType) {
+			return new StringType();
+		}
+
+		return $defaultReturnType;
+	}
+
+	/**
+	 * bcsqrt
+	 * https://www.php.net/manual/en/function.bcsqrt.php
+	 * > Returns the square root as a string, or NULL if operand is negative.
+	 *
+	 * @param FuncCall $functionCall
+	 * @param Scope $scope
+	 * @return NullType|StringType|UnionType
+	 */
+	private function getTypeForBcSqrt(FuncCall $functionCall, Scope $scope)
+	{
+		$defaultReturnType = new UnionType([new StringType(), new NullType()]);
+
+		if (isset($functionCall->args[0]) === false) {
+			return $defaultReturnType;
+		}
+
+		$firstArgument = $scope->getType($functionCall->args[0]->value);
+
+		if ($firstArgument instanceof UnaryMinus ||
+			($firstArgument instanceof ConstantStringType && $firstArgument->getValue() < 0)) {
+			return new NullType();
+		}
+
+		if ($scope->getType($functionCall->args[0]->value) instanceof ConstantStringType) {
+			return new StringType();
+		}
+
+		return $defaultReturnType;
+	}
+
+	/**
+	 * bcpowmod()
+	 * https://www.php.net/manual/en/function.bcpowmod.php
+	 * > Returns the result as a string, or FALSE if modulus is 0 or exponent is negative.
+	 * @param FuncCall $functionCall
+	 * @param Scope $scope
+	 * @return BooleanType|StringType|UnionType
+	 */
+	private function getTypeForBcPowMod(FuncCall $functionCall, Scope $scope)
+	{
+		if (isset($functionCall->args[1]) === false) {
+			return new UnionType([new StringType(), new BooleanType()]);
+		}
+
+		$exponent = $scope->getType($functionCall->args[1]->value);
+
+		if ($exponent instanceof UnaryMinus || ($exponent instanceof ConstantStringType && $exponent->getValue() < 0)) {
+			return new BooleanType(); // false
```

Comment:
This is always boolean `false`. Can I return this somehow?

Reply:
`new ConstantBooleanType(false)`